core/unit_test/TestGraph.hpp

```diff
@@ -0,0 +1,242 @@
+        count_functor{count, bugs, 0, 6});
+    //----------------------------------------
+    ready.then_parallel_for(Kokkos::TeamPolicy<TEST_EXECSPACE>{1, 1},
+                            count_functor{count, bugs, 0, 6});
+    //----------------------------------------
+    ready.then_parallel_for(2, count_functor{count, bugs, 0, 6});
+    //----------------------------------------
+  });
+  graph.submit();
+  Kokkos::deep_copy(ex, count_host, count);
+  Kokkos::deep_copy(ex, bugs_host, bugs);
+  ex.fence();
+
+  ASSERT_EQ(6, count_host());
+  ASSERT_EQ(0, bugs_host());
+}
+
+TEST_F(TEST_CATEGORY_FIXTURE(count_bugs), when_all_cycle) {
+  view_type reduction_out{"reduction_out"};
+  view_host reduction_host{"reduction_host"};
+  Kokkos::Experimental::create_graph(ex, [=](auto root) {
+    //----------------------------------------
+    // Test when_all when redundant dependencies are given
+    auto f1 = root.then_parallel_for(1, set_functor{count, 0});
+    auto f2 = f1.then_parallel_for(1, count_functor{count, bugs, 0, 0});
+    auto f3 = f2.then_parallel_for(5, count_functor{count, bugs, 1, 5});
+    auto f4 = Kokkos::Experimental::when_all(f2, f3).then_parallel_for(
+        1, count_functor{count, bugs, 6, 6});
+    Kokkos::Experimental::when_all(f1, f4, f3)
+        .then_parallel_reduce(6, set_result_functor{count}, reduction_out);
+    //----------------------------------------
+  }).submit();
+  Kokkos::deep_copy(ex, bugs_host, bugs);
+  Kokkos::deep_copy(ex, count_host, count);
+  Kokkos::deep_copy(ex, reduction_host, reduction_out);
+  ex.fence();
+  ASSERT_EQ(0, bugs_host());
+  ASSERT_EQ(7, count_host());
+  ASSERT_EQ(42, reduction_host());
+  //----------------------------------------
+}
+
+// This test is disabled because we don't currently support copying to host,
+// even asynchronously. We _may_ want to do that eventually?
+TEST_F(TEST_CATEGORY_FIXTURE(count_bugs), DISABLED_repeat_chain) {
+  auto graph = Kokkos::Experimental::create_graph(
+      ex, [=, count_host = count_host](auto root) {
+        //----------------------------------------
+        root.then_parallel_for(1, set_functor{count, 0})
+            .then_parallel_for(1, count_functor{count, bugs, 0, 0})
+            .then_parallel_for(1, count_functor{count, bugs, 1, 1})
+            .then_parallel_reduce(1, set_result_functor{count}, count_host)
+            .then_parallel_reduce(
+                1, set_result_functor{bugs},
+                Kokkos::Sum<int, Kokkos::HostSpace>{bugs_host});
+        //----------------------------------------
+      });
+
+  //----------------------------------------
+  constexpr int repeats = 10;
+
+  for (int i = 0; i < repeats; ++i) {
+    graph.submit();
+    ex.fence();
+    EXPECT_EQ(2, count_host());
+    EXPECT_EQ(0, bugs_host());
+  }
+  //----------------------------------------
+}
+
+// This won't work with the way we're currently doing things
```

Comment:
Why exactly is this not working?

Reply:
This used to not work because of the zero-size reduction bug, but I guess that's fixed now?

Reply:
Resolved by 8d8f0564

Reply:
ah makes sense. Yeah if this is rebased on top of the stuff which solved the zero size reduction thing we should be good.